server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -871,6 +883,16 @@ private void verifyRelocatingState() {
             throw new IllegalIndexShardStateException(shardId, IndexShardState.STARTED, ": shard is no longer relocating " + shardRouting);
         }
 
+        if (targetNodeId != null) {
+            if (targetNodeId.equals(shardRouting.relocatingNodeId()) == false) {
+                throw new IllegalIndexShardStateException(
+                    shardId,
+                    IndexShardState.STARTED,
+                    ": shard is no longer relocating to node [" + targetNodeId + "]: " + shardRouting
```

Comment:
nit: this is implied but should it be mentioned that the primary relocation got cancelled?

Reply:
I think saying that might raise more questions than it answers - there isn't really a user-facing notion of cancelling a recovery, it's just something that happens behind the scenes.